src/Components/Server/src/Authentication/CircuitAuthenticationHandler.cs

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Components.Server.Circuits;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Components.Server
+{
+    // Authentication handler for circuit ids.
+    // We pair each circuit id with a cookie.
+    // We use an authentication handler to enable smooth integration with signalr service, as it preserves limited
+    // information about the connection context (amongs it the Principal), that's why we do it that way.
+    internal class CircuitAuthenticationHandler : SignInAuthenticationHandler<CircuitAuthenticationOptions>
```

Comment:
This is the weirdest authentication handler I have ever seen.
- No challenge.
- Signin ignores the user
- No signout

How exactly does this "enable smooth integration with signalr service"?

How well does this stack when signalr really is using auth?

Reply:
> This is the weirdest authentication handler I have ever seen.
> 
> * No challenge.

What would the challenge would be? Simply return a 401?

> * Signin ignores the user

We are not authenticating a specific user but a specific circuit id. I could have build a principal and passed in the properties through claims on it. That might be more "correct".

> * No signout

We might implement that if we end up offering an endpoint to close the circuit.


> How exactly does this "enable smooth integration with signalr service"?

The information flows transparently through the principal whether you are using signalr service or not. Signalr service copies a limited amount of information, if we were to do something different it would require them to do something specific for blazor, which we want to avoid.

> How well does this stack when signalr really is using auth?

It stacks fine. There are two identities, one for whatever auth system you use and another one for the connection. When using the singnalr service all identities get merged. The hub authentication is based on a policy and by default it gets configured to use the default scheme + our scheme, resulting in what you have configured + our scheme. Its also available for you to tweak it to your heart's contents.

Reply:
For the record, I'm killing the auth hander as I found a better approach